src/main/java/org/opensearch/flowframework/workflow/CreateIndex/CreateIndexStep.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+package org.opensearch.flowframework.workflow.CreateIndex;
+
+import com.google.common.base.Charsets;
+import com.google.common.io.Resources;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.opensearch.action.ActionListener;
+import org.opensearch.action.admin.indices.create.CreateIndexRequest;
+import org.opensearch.action.admin.indices.create.CreateIndexResponse;
+import org.opensearch.client.Client;
+import org.opensearch.common.xcontent.XContentType;
+import org.opensearch.flowframework.workflow.WorkflowData;
+import org.opensearch.flowframework.workflow.WorkflowStep;
+
+import java.io.IOException;
+import java.net.URL;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+
+public class CreateIndexStep implements WorkflowStep {
+
+    private static final Logger logger = LogManager.getLogger(CreateIndexStep.class);
+    private Client client;
+    private final String CREATE_INDEX_STEP = "create_index_step";
+
+    public CreateIndexStep(Client client) {
+        this.client = client;
+    }
+
+    @Override
+    public CompletableFuture<WorkflowData> execute(List<WorkflowData> data) {
+        CompletableFuture<WorkflowData> future = new CompletableFuture<>();
+        ActionListener<CreateIndexResponse> actionListener = new ActionListener<>() {
+
+            @Override
+            public void onResponse(CreateIndexResponse createIndexResponse) {
+                logger.info("created index:{}");
+                future.complete(new WorkflowData() {
+                    @Override
+                    public Map<String, Object> getContent() {
+                        return Map.of("index", createIndexResponse.index());
+                    }
+                });
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                logger.error("Index creation failed", e);
+                future.completeExceptionally(e);
+            }
+        };
+
+        String index = null;
+
+        for (WorkflowData workflowData : data) {
+            // Fetch index from content i.e. request body of execute API
+            Map<String, Object> content = workflowData.getContent();
+            index = (String) content.get("index");
```

Comment:
As written, this will iterate all of the previous steps' data, with later ones overwriting earlier ones.  This may not be what you want.

The first element in `data` will be the initialization from the template parsing, most likely to contain any user-provided info.  The later elements are any immediate predecessor steps but ordering is not guaranteed.

TLDR: you might want a 'break;` here if index is not `null`.